src/TheSpace/HarbergerMarket.sol

```diff
@@ -323,81 +226,71 @@ contract HarbergerMarket is ERC721Enumerable, IHarbergerMarket, Multicall, ACLMa
-        treasuryRecord.accumulatedTreasury += treasuryAdded;
-
-        // update accumulated ubi
-        treasuryRecord.accumulatedUBI += (amount - treasuryAdded);
-
-        // update tax record
-        tokenRecord[tokenId_].lastTaxCollection = block.number;
+        // calculate treasury change
+        uint256 treasuryShare = registry.taxConfig(IHarbergerRegistry.ConfigOptions.treasuryShare);
+        uint256 treasuryAdded = (amount_ * treasuryShare) / 10000;
+
+        // set treasury record
+        (uint256 accumulatedUBI, uint256 accumulatedTreasury, uint256 treasuryWithdrawn) = registry.treasuryRecord();
+        registry.setTreasuryRecord(
+            accumulatedUBI + (amount_ - treasuryAdded),
+            accumulatedTreasury + treasuryAdded,
+            treasuryWithdrawn
+        );
+
+        // update lastTaxCollection and emit tax event
+        (uint256 price, , uint256 ubiWithdrawn) = registry.tokenRecord(tokenId_);
+        registry.setTokenRecord(tokenId_, price, block.number, ubiWithdrawn);
+        registry.emitTax(tokenId_, taxpayer_, amount_);
+    }
 
-        emit Tax(tokenId_, taxpayer, amount);
+    /// @inheritdoc IHarbergerMarket
+    function settleTax(uint256 tokenId_) public returns (bool success) {
+        success = _collectTax(tokenId_);
+        if (!success) registry.burn(tokenId_);
     }
 
     /// @inheritdoc IHarbergerMarket
     function ubiAvailable(uint256 tokenId_) public view returns (uint256) {
-        return treasuryRecord.accumulatedUBI / _totalSupply - tokenRecord[tokenId_].ubiWithdrawn;
+        (uint256 accumulatedUBI, , ) = registry.treasuryRecord();
+        (, , uint256 ubiWithdrawn) = registry.tokenRecord(tokenId_);
+
+        return accumulatedUBI / registry.totalSupply() - ubiWithdrawn;
     }
 
     /**
      * @notice Withdraw UBI on given token.
      */
     function withdrawUbi(uint256 tokenId_) external {
-        uint256 ubi = ubiAvailable(tokenId_);
+        uint256 amount = ubiAvailable(tokenId_);
 
-        if (ubi > 0) {
-            tokenRecord[tokenId_].ubiWithdrawn += ubi;
+        if (amount > 0) {
+            // transfer
+            address recipient = registry.ownerOf(tokenId_);
+            registry.transferCurrency(recipient, amount);
 
-            address recipient = ownerOf(tokenId_);
-            currency.transfer(recipient, ubi);
+            // record
+            (uint256 price, uint256 lastTaxCollection, uint256 ubiWithdrawn) = registry.tokenRecord(tokenId_);
+            registry.setTokenRecord(tokenId_, price, lastTaxCollection, ubiWithdrawn + amount);
 
-            emit UBI(tokenId_, recipient, ubi);
+            // emit event
+            registry.emitUBI(tokenId_, recipient, amount);
         }
     }
-
-    /**
-     * @notice Internel function to set price for a token.
-     */
-    function _setPrice(uint256 tokenId_, uint256 price_) internal {
-        _setPrice(tokenId_, price_, ownerOf(tokenId_));
-    }
-
-    function _setPrice(
```

Comment:
overloading `_setPrice` with three params is for gas optimization (skip reading `ownerOf(tokenId_)`).